### ISSUE TITLE
Save Volume Settings

### ISSUE THREAD

**Issue by SpikeyMikey** [kyleneideck/BackgroundMusic#180]
Is there a way to save volume settings for apps? I'm liking this audio utility, but every time I restart my MAC and/or quit-relaunch programs, the Background Music volume settings reset all apps to the zero-default position. It's getting annoying that I have to repetitively adjust the sliders and BM doesn't remember the last volume settings from the prior times the apps were open.

**Comment by dresnick-sf** [comment]
I actually only use the app to help me with lowering the volume, for when I use noise isolating headphones. So keeping my custom volume levels after a restart is a big feature I'd really like to see too.

**Comment by zxzzz8** [comment]
I started using this app recently as well just for the volume control it have, and i really like it, but not having the app saving the volume setting for each app, is pretty detrimental for me, it feels like a really critical feature....

**Comment by RichSimon** [comment]
For a very related feature, would be fantastic to save multiple configurations, and also to say "mute all other apps". If I go on a telecon, ticking the Webex configuration that mutes everything aside from Webex would be :)

**Comment by ZappoB** [comment]
Since this request is open and unanswered by the author for over 10 months, it seems very unlikely to show up in the near future, so I suppose.

It would be great, if the author at least leave a statement. No complaining, it's a great tool and free, too.

**Comment by kyleneideck** [comment]
@ZappoB I agree that it's an important feature. It's been on my to do list for a long time now. I don't know when I'll get around to it.

**Comment by CrystalSpore** [comment]
Just wanted throw my hat in here. It would be pretty great to have this so that I don't blast my ears out with Spotify whenever I restart my machine : )

**Comment by kyleneideck** [comment]
@CrystalSpore When a process registers with CoreAudio to play audio, CoreAudio sends BGMDriver (the virtual audio driver) the process's bundle ID. CoreAudio calls the process a "client". When you change an app's volume, BGMApp sends the app's bundle ID to BGMDriver and BGMDriver tries to match it to a client.

To save the app volumes (and pans), I think we would need to save [`mBundleID`](https://github.com/kyleneideck/BackgroundMusic/blob/master/BGMDriver/BGMDriver/DeviceClients/BGM_Client.h#L61), [`mRelativeVolume`](https://github.com/kyleneideck/BackgroundMusic/blob/master/BGMDriver/BGMDriver/DeviceClients/BGM_Client.h#L73) and [`mPanPosition`](https://github.com/kyleneideck/BackgroundMusic/blob/master/BGMDriver/BGMDriver/DeviceClients/BGM_Client.h#L76) for each client that has a volume or pan set in BGMDriver.

Another option would be to save the app volumes in BGMApp and send them to BGMDriver when BGMApp starts up. It's been a while since I last looked at this code, but it looks like [`setVolume`](https://github.com/kyleneideck/BackgroundMusic/blob/master/BGMApp/BGMApp/BGMAppVolumesController.mm#L160) and [`setPanPosition`](https://github.com/kyleneideck/BackgroundMusic/blob/master/BGMApp/BGMApp/BGMAppVolumesController.mm#L199) in `BGMAppVolumesController` would be good places to save the volumes/pans.

BGMApp currently [gets the app volumes from BGMDriver](https://github.com/kyleneideck/BackgroundMusic/blob/ffa86bbcd98133ffd5dbcc2e7585b6af57030677/BGMApp/BGMApp/BGMAppVolumesController.mm#L94) when it starts up, so I think it would need to get them from user defaults instead and send them to BGMDriver.

BGMApp already saves some user data -- see [`BGMUserDefaults.h`](https://github.com/kyleneideck/BackgroundMusic/blob/master/BGMApp/BGMApp/BGMUserDefaults.h) -- so it might be easier to save the app volumes in the same way. On the other hand, it sounds like saving the volumes in BGMDriver might end up being a smaller change.

I don't think it really matters from a user's perspective whether they're saved in BGMDriver or BGMApp, so I'd recommend picking the option you think will be easier.

**Comment by synclpz** [comment]
You should save it in BGMDriver or periodically (or on event?) transmit volume/pan settings from app to driver, because if an app is started after saved settings are transmitted/applied the app will use default. I tried to use BGM to mute sounds from annoying apps (like corporate VPN client) but no luck since it does not remember settings between BGM or VPN app restarts :(

**Comment by synclpz** [comment]
Same issue #331

**Comment by shane-deveire** [comment]
I know this project is pretty much unsupported at this stage but please do add this feature if you do another release.

**Comment by jwshive** [comment]
Came here specifically to find this enhancement request. Would be a fantastic feature.
